src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseReport.java

```diff
@@ -0,0 +1,607 @@
+    rowNum -= (thisMonthTwelve.size() + 4); // list + (blank, 2 headers, total)
+
+    row = maybeInList(TimePoint.EIGHTEEN.getKey(), labelListShort, thisMonthEighteen, rowNum);
+    if (row != null) return row;
+    rowNum -= (thisMonthEighteen.size() + 4); // list + (blank, 2 headers, total)
+
+    row = maybeInList(TimePoint.TWENTY_FOUR.getKey(), labelListLong, thisMonthTwentyFour, rowNum);
+    if (row != null) return row;
+    rowNum -= (thisMonthTwentyFour.size() + 4); // list + (blank, 2 headers, total)
+
+    row = maybeInList("Tissue", labelListSlides, thisMonthSlides, rowNum);
+    if (row != null) return row;
+    rowNum -= (thisMonthSlides.size() + 4); // list + (blank, 2 headers, total)
+
+    if (rowNum == 0) {
+      return makeBlankRow();
+    } else if (rowNum == 1) {
+      return makeAlmostBlankRow(String.format("Total Inventory Available (%s)", new SimpleDateFormat(DATE_FORMAT).format(new Date())));
+    }
+    rowNum -= 2;
+
+    row = maybeInList("Total in Inventory", labelListInventory, inventoryAvailable, rowNum);
+    if (row != null) return row;
+    rowNum -= (inventoryAvailable.size() + 4); // list + (blank, 2 headers, total)
+
+    if (rowNum == 0) {
+      return makeBlankRow();
+    } else if (rowNum == 1) {
+      String[] blank = makeBlankRow();
+      blank[0] = "Samples with mismatched alias and cryovial/slide number";
+      return blank;
+    }
+
+    row = maybeSampleAlias(barcodeAndNameCodingMismatch, rowNum);
+    if (row != null) return row;
+    rowNum -= barcodeAndNameCodingMismatch.size();
+
+    if (rowNum == 0) {
+      return makeBlankRow();
+    } else if (rowNum == 1) {
+      String[] blank = makeBlankRow();
+      blank[0] = "Slides with non-slide timepoint in alias";
+      return blank;
+    }
+
+    row = maybeSampleAlias(slidesWithoutSlideTimepoint, rowNum);
+    if (row != null) return row;
+    rowNum -= slidesWithoutSlideTimepoint.size();
+
+    return makeBlankRow();
+  }
+
+  private String[] maybeInList(String category, List<SampleLabel> labels, List<List<String>> list, int rowNum) {
+    if (rowNum == 0) {
+      return makeBlankRow();
+    } else if (rowNum == 1) {
+      return generateRow(labelListHeader(category, labels));
+    } else if (rowNum == 2) {
+      return generateRow(sampleCasesHeader(labels));
+    }
+    rowNum -= 3;
+    if (rowNum < list.size()) {
+      return generateRow(list.get(rowNum));
+    } else if (rowNum == list.size()) {
+      return generateTotalsRow(list, labels);
+    }
+    return null;
+    // do our subtraction of the list + totals outside this method, because changes here won't affect rowNum outside the method
+  }
+
+  private String[] maybeSampleAlias(List<SampleDto> samples, int rowNum) {
+    if (rowNum < samples.size()) {
+      SampleDto sample = samples.get(rowNum);
+      String[] row = makeAlmostBlankRow(sample.getName());
+      row[1] = sample.getTubeBarcode();
+      return row;
+    }
+    return null;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)